Decouple flask app

btylerburton opened the pull request:
# Pull Request

Related to https://github.com/GSA/data.gov/issues/4865

## About

Allows us to run follow-up jobs to sync records which may have previously errored out due network side-effects.

Moves towards ability to run harvest clear and records validation as first-class jobs.

EDIT: This PR is a beast. Make sure to expand the changes to `harvester/harvest.py`. I recommend looking at those in the side-by-side editor as it's a thorough rewrite.

## PR TASKS

- [X] The actual code changes.
- [X] Tests written and passed.
- [x] Any changes to docs?

btylerburton committed:
cleans up routes and makes facets more flexible

btylerburton commented:
~DRAFT WIP: not finished yet... but wanted to put this up as we're pausing on harvester work for a sprint and having eyes on this could help clean it up...~

btylerburton committed:
updated sequence

btylerburton committed:
cleanup and fix tests

btylerburton committed:
lint

btylerburton committed:
update ruff rules

btylerburton committed:
passing tests; plus new docs

btylerburton committed:
skip deploy action until we're ready

btylerburton committed:
lint

btylerburton committed:
incorporate and test different job flows

btylerburton committed:
revert ckan_id to string

btylerburton committed:
comments

btylerburton committed:
pass job_type into load_manager start

btylerburton committed:
import fixes

rshewitt commented on .github/workflows/deploy.yml:
why do we want to turn off deployment?

btylerburton replied:
We don't have apps in staging or prod yet

btylerburton committed:
fixes from testing

rshewitt commented:
a lot has changed here and looks like pytest coverage is [broken](https://github.com/GSA/datagov-harvester/actions/runs/12918107080/job/36025960186?pr=115#annotation:7:31) (and has been broken for a bit). can this be fixed in this PR? low coverage could be fixed in a separate ticket.

btylerburton committed:
Revert "Merge pull request #126 from GSA/update-cf-client"

This reverts commit 1e7cd5d35ef51020819183510733df7e682062ee, reversing
changes made to 1718c3a0e523b68bd10e1e1ce678f6a5f2adc6c6.

btylerburton commented:
> a lot has changed here and looks like pytest coverage is [broken](https://github.com/GSA/datagov-harvester/actions/runs/12918107080/job/36025960186?pr=115#annotation:7:31) (and has been broken for a bit). can this be fixed in this PR? low coverage could be fixed in a separate ticket.

I didnt realize that coverage action depended on the xml. I can update that.

btylerburton committed:
fix test coverage reporter action

jbrown-xentity commented on docs/diagrams/mermaid/dest/job_follow_up_job-1.svg:
We can probably remove Agency Source and MDTranslator from this diagram, they aren't used and make things more busy.

jbrown-xentity commented on docs/diagrams/mermaid/dest/job_full_harvest-1.svg:
I think there's a missing piece in the LOAD step; that should always save in the DB that the record is loaded into CKAN, right? That way the follow up job doesn't have to analyze, it will just pick up where it left off?

btylerburton replied:
docs should all be updated with that info.

jbrown-xentity commented on docs/diagrams/mermaid/dest/job_restart_job-1.svg:
Do we have documentation on a restarted job vs a follow-up job? They seem almost identical...

btylerburton replied:
refactored to just a single sync job after a chat with reid.

jbrown-xentity reviewed (commented):
This is a review of the diagrams.
@btylerburton and I agreed that the historical diagrams don't make sense here, as they are overly thorough and confusing. These will be moved to a google drive folder, with a link for posterity.
The other comments are hopefully helpful to improving and clarifying the diagrams for future users.
@cmhedrickREI I encourage you to look at the new ones, as they show the high level systems, what is talking with what, and when.

btylerburton committed:
fixes request args type casting issue;
makes paginate decorator cleaner;
adds harvest clear helper to clean up records in error;
TODO: figure out why clear helper has dirty records in a test system

rshewitt commented on .github/workflows/commit.yml:
```diff
@@ -51,8 +51,10 @@ jobs:
         with:
-          pytest-coverage-path: pytest-coverage.txt
-          junitxml-path: pytest.xml
+          multiple-files: |
```
coverages still don't appear to be [working](https://github.com/GSA/datagov-harvester/actions/runs/12944490684/job/36105474355#step:7:1). it's not a huge deal. can create a separate ticket for it.

btylerburton replied:
i'm looking into that now. just pushed up a test commit.

rshewitt commented on app/routes.py:
```diff
@@ -465,23 +465,22 @@ def view_harvest_source_data(source_id: str):
-        skip_pagination=True,
         source_id=source.id,
     )
     ckan_records_count = db.get_harvest_records_by_source(
         count=True,
-        skip_pagination=True,
         source_id=source.id,
-        facets=["ckan_id is not null"],
+        facets="ckan_id is not null",
     )
     error_records_count = db.get_harvest_records_by_source(
         count=True,
-        skip_pagination=True,
         source_id=source.id,
-        facets=["status = 'error'"],
+        facets="status = 'error'",
+    )
```
getting counts from `db.get_harvest_records_by_source` would include historic records ( e.g. record_count and ckan_records_count = 2 given a create then update ). error_records_count could also return more than expected if a record failed validation, is fixed by the data provider, then fails sync. here's the [get_latest_records](https://github.com/GSA/datagov-harvester/blob/1e7cd5d35ef51020819183510733df7e682062ee/database/interface.py#L420) function. it could be useful in these cases?

btylerburton replied:
i didn't add this block, just cleaned it up, but i agree we should not be showing "shadow" records that are deleted; this is only useful for the internal harvester when clearing a source. it took me a long time to reconcile exactly what that number meant.

rshewitt commented on harvester/harvest.py:
```diff
@@ -104,17 +112,34 @@ def __post_init__(self) -> None:
             self.dataset_schema = open_json(
                 ROOT_DIR / "schemas" / "non-federal_dataset.json"
             )
+        self._validator = Draft202012Validator(self.dataset_schema)
```
+1. no need to create this every time we validate a record.

rshewitt commented on harvester/harvest.py:
```diff
@@ -197,37 +222,38 @@ def external_records_to_id_hash(self, records: list[dict]) -> None:
-                if job.job_type == 'clear':
-                    self.external_records_to_id_hash([])
-                else:
+            if job.job_type == "clear":
```
this method is called in `extract`. why does the clear operation need an extraction to occur? putting this here because this prompted my discovery

btylerburton replied:
clear works by diffing against an empty incoming source list. i wonder if we can move that insertion up the chain, though. will take a look.

btylerburton committed:
updates to docs;
delete harvest record now deletes all instances of a record;
explicitly close the db connection after a harvest;
clean up README;
adds steps to document harvesting to local CKAN

rshewitt commented on database/interface.py:
```diff
@@ -237,6 +241,17 @@ def get_first_harvest_jobs_by_filter(self, filter):
+            self.db.query(HarvestJob)
+            .filter_by(harvest_source_id=source_id)
+            .order_by(HarvestJob.date_created.desc())
+            .limit(2)
```
so follow-up jobs are supposed to be responsible for completing unfinished work from a previous harvest job which can be represented by (work completed) - (work to-be-done). “follow_up_job_helper“ expects the job report to have occurred in order to determine the work completed. What kind of record-level work would need to be followed up on? We wouldn’t be following up on transformation or validation error records because we’re using the same data as last time so we’d get the same result.

btylerburton replied:
we would be following up on a broken sync. we've definitely talked about a way to force a sync in isolation. perhaps i should just pare back `follow-up` and `restart` and just make a `sync` job?

rshewitt replied:
i think that would be simpler. do we want to restart a job older than what follow-up already does? thinking about where restart could come in where follow-up doesn't

btylerburton replied:
ok, i'm going to pare this back to a simple `sync` job.

btylerburton committed:
lint

btylerburton committed:
update coverage report

btylerburton committed:
push with app context

rshewitt commented on .github/workflows/commit.yml:
```diff
@@ -64,12 +77,16 @@ jobs:
-        uses: actions/setup-python@v5
-        with:
-          python-version: ${{ env.PY_VERSION }}
+          node-version: "20"
```
you should be able to use the package.json file and it will discover the engine version

btylerburton replied:
next pr should test a new config based on these docs: https://github.com/actions/setup-node/blob/main/docs/advanced-usage.md#working-with-lockfiles

rshewitt commented on harvester/harvest.py:
```diff
@@ -275,173 +300,250 @@ def write_compare_to_db(self) -> dict:
+                records.append(record)
+
+        # set records on new
+        self._records = records
```
_records shouldn't be set directly here. you'll want to add a setter where the "records" property is declared.

rshewitt commented on harvester/harvest.py:
```diff
@@ -275,173 +300,250 @@ def write_compare_to_db(self) -> dict:
+                )  # don't allow for negative values based on previous results
+
         self.db_interface.update_harvest_job(self.job_id, job_status)
+        self._report = job_status
```
same here. _report attr shouldn't be set explicitly.

btylerburton replied:
resolved

rshewitt commented on harvester/harvest.py:
```diff
@@ -275,173 +300,250 @@ def write_compare_to_db(self) -> dict:
                     )
                     logger.info(f"Notification email sent to: {recipient}")
+
         except Exception as e:
```
not sure how frequently this would occur but does only logging this give us the visibility we're looking for? would we want to create a custom exception and add it as a job error?

btylerburton replied:
yes! i'll add a todo and we can take care of this in another ticket.

rshewitt commented on harvester/harvest.py:
```diff
@@ -275,173 +300,250 @@ def write_compare_to_db(self) -> dict:
+        records = []
+        for db_record in db_records:
+            records.append(self.make_record_contract(db_record, input_type="orm"))
+        self._records.extend(records)
```
same. update to self.records not self._records

btylerburton replied:
resolved

btylerburton committed:
updates to setters for records and report;
removes follow-up and restart jobs in favor of sync;
adds validate job tests;
adds records_validated to harvest_job model;
other responses to feedback

btylerburton committed:
Merge branch 'main' of github.com:GSA/datagov-harvester

btylerburton committed:
break up extract and compare; fixes builder query facets issue

btylerburton committed:
updates to docs

btylerburton committed:
Reapply "Merge pull request #126 from GSA/update-cf-client"

This reverts commit 1c2b6275024a1071c623345b7e5b15048a093e5e.

btylerburton committed:
update gh workflows

rshewitt commented on database/interface.py:
```diff
@@ -237,6 +238,17 @@ def get_first_harvest_jobs_by_filter(self, filter):
         )
         return harvest_job
 
+    def get_harvest_jobs_by_source_id(self, source_id):
```
looks like this function isn't being used anymore

btylerburton replied:
Probably. Lots of cleanup still needed.

rshewitt commented on harvester/harvest.py:
```diff
@@ -275,173 +302,240 @@ def write_compare_to_db(self) -> dict:
+        }
+        self._report = results
+        new_records = []
+        for job_record in job.records:
```
we should be skipping records with a status = "success"

btylerburton replied:
we skip those in the `record.sync()`, but it might be good to exclude those here.

btylerburton replied:
updated just in case and no tests failed

btylerburton committed:
don't try to resync records with status:success

btylerburton committed:
Merge branch 'main' of github.com:GSA/datagov-harvester

rshewitt reviewed (approved):
we got a lotta good work over here!
